### ISSUE TITLE
Hard loss 有用吗？

### ISSUE THREAD

**Issue by sz128** [airaria/TextBrewer#39]
作者，您有试过 加hard loss的实验吗？

如果student模型是随机初始化的话，我在cmrc2018的实验中尝试添加了 hard loss且hard_loss_weight=1的时候蒸馏训练结果奇差（F1 25%左右，EM 7% 左右）。即便，我将hard_loss_weight设置成0.001这样超级小的值，也会让原本蒸馏最好的性能下降0.5%-1% 大概。

我非常疑惑“添加hard loss”为什么会有这么明显的伤害？期待解答。

**Comment by airaria** [comment]
我在cmrc2018上尝试hard loss 没什么用，但不至于掉这么多，最差也应该能达到直接训练的水平。
你的实验结果似乎不太正常。检查一下student adaptor返回的字典中的'losses'是否正确匹配上了模型返回的loss？

**Comment by sz128** [comment]
我检查了一下，确认没有问题。

确实像你说的，最差也应该能到达直接训练的水平。
1）如果student模型是采用预训练的语言模型（比如electra-small), 则加上hard_loss挺正常的。
2）如果student模型是完全随机初始化的，则加上hard_loss会明显把训练过程往 “随机初始化模型+直接训练”的方向带（就是我前面说到hard_loss_weight=1的时候蒸馏结果只有F1 25%左右、EM 7% 左右的水平）。这种情况如何解释呢？

**Comment by airaria** [comment]
> 我检查了一下，确认没有问题。
> 
> 确实像你说的，最差也应该能到达直接训练的水平。
> 1）如果student模型是采用预训练的语言模型（比如electra-small), 则加上hard_loss挺正常的。
> 2）如果student模型是完全随机初始化的，则加上hard_loss会明显把训练过程往 “随机初始化模型+直接训练”的方向带（就是我前面说到hard_loss_weight=1的时候蒸馏结果只有F1 25%左右、EM 7% 左右的水平）。这种情况如何解释呢？

在模型是随机初始化的情况下，这种情况我之前确实没遇见过，也无法解释。猜测的话估计是soft-label信息量更大吧
如果不是代码有bug，那是个蛮有趣的现象

**Comment by sxthunder** [comment]
> 模型是采用预训练的语言模型（比如electra-small), 则加上hard_loss挺正常的。

理论上soft loss和hard loss一起训练时，soft loss应该乘上 T^2来保证soft和hard的梯度是同量级的，但是在源码中我并没有找到类似的代码。是否因为这个原因导致soft loss在hard loss下几乎不起作用了？

**Comment by sz128** [comment]
@sxthunder 您说的T是指temperature吗？我后续已经没有做过相关试验了。我记得我试过把temperature设为1进行实验，结果依旧不理想。

**Comment by catqaq** [comment]
@sz128 @airaria Hi，我这儿遇到了类似的问题。在hiddens等中间层特征的基础上，尝试引入hard loss，发现效果变差了。起初，我以为是权重设置不合理，但在大量尝试各种权重组合之后，应该可以排除这个可能性。然后，我试图只用hard loss进行训练（相当于借了textbrewer的壳子在做直接训练），结果发现直接训练的效果也很差，而且是越学越差，比起点还差。
adaptor定义大概像这样：
` def simple_adaptor(batch, model_outputs):
        loss = simcse_loss(model_outputs["se"])
        return {
		    "losses": [loss],
            # 'hidden': model_outputs["hiddens"],
            # 'attention': model_outputs["attentions"],
        }`

**Comment by sz128** [comment]
@catqaq 当时我的实验也是存在这个问题。不过我后来没有继续深入去寻找原因。

**Comment by catqaq** [comment]
@sz128 嗯，和你遇到的情况一样，如果模型已经训练过了，则只用hard loss或hard  loss+蒸馏都会比较正常，但是如果从头训练，hard loss就会有问题。

**Comment by airaria** [comment]
> @sz128 嗯，和你遇到的情况一样，如果模型已经训练过了，则只用hard loss或hard loss+蒸馏都会比较正常，但是如果从头训练，hard loss就会有问题。

如果只用textbrewer的壳子训练，用textbrewer.BasicTrainer试试？
我另外检查一下hard_loss的实现有没有问题

**Comment by catqaq** [comment]
@airaria 您好，感谢建议。有空我会再验证下**BasicTrainer中直接训练hard loss**和**在GeneralDistiller中训练只保留hard loss**的差异。
不过，事实上我们并不需要用textbrewer的壳子来训练，而是在蒸馏时发现引入hard loss后效果变差了，才想着单独验证一下hard loss的影响，于是在蒸馏时只保留了hard loss，然后就发现了前述的诡异现象。

**Comment by catqaq** [comment]
补充：刚才看了下代码，感觉BasicTrainer中直接训练hard loss和在GeneralDistiller中训练只保留hard loss似乎没啥差异：
loss都是从adaptor中返回的，后续的处理也都差不多；
对于GeneralDistiller中训练只保留hard loss：
`results_S = post_adaptor(self.adaptor_S(student_batch,results_S))
for loss in results_S['losses']:
    # in case of multi-GPU
    total_hl_loss += loss.mean() `
在BasicTrainer中直接训练hard loss：
`results = post_adaptor(self.adaptor(batch,results))
for loss in results['losses']:
    # in case of multi-GPU
    total_loss += loss.mean()`
